pkg/azure/validation/validation_test.go

```diff
@@ -0,0 +1,477 @@
+		},
+		{"should forbid empty subnetName and vnetName",
+			"", "", 2,
+			ConsistOf(
+				PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("providerSpec.subnetInfo.vnetName")})),
+				PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("providerSpec.subnetInfo.subnetName")})),
+			),
+		},
+		{"should succeed when vnetName and subnetName are present", testVnetName, testSubnetName, 0, nil},
+	}
+	g := NewWithT(t)
+	for _, entry := range table {
+		t.Run(entry.description, func(t *testing.T) {
+			subnetInfo := api.AzureSubnetInfo{
+				VnetName:   entry.vnetName,
+				SubnetName: entry.subnetName,
+			}
+			errList := validateSubnetInfo(subnetInfo, fldPath)
+			g.Expect(len(errList)).To(Equal(entry.expectedErrors))
+			if entry.matcher != nil {
+				g.Expect(errList).To(entry.matcher)
+			}
+		})
+	}
+
+}
+
+func TestValidateHardwareProfile(t *testing.T) {
+	fldPath := field.NewPath("providerSpec.properties.hardwareProfile")
+	hwProfile := api.AzureHardwareProfile{}
+	g := NewWithT(t)
+	errList := validateHardwareProfile(hwProfile, fldPath)
+	g.Expect(len(errList)).To(Equal(1))
+	g.Expect(errList).To(ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("providerSpec.properties.hardwareProfile.vmSize")}))))
+}
+
+func TestValidateOSDisk(t *testing.T) {
+	fldPath := field.NewPath("providerSpec.properties.storageProfile.osDisk")
+	table := []struct {
+		description    string
+		osDisk         api.AzureOSDisk
+		expectedErrors int
+		matcher        gomegatypes.GomegaMatcher
+	}{
+		{
+			"should forbid empty createOption",
+			api.AzureOSDisk{Name: "osdisk-0", DiskSizeGB: 20, CreateOption: ""}, 1,
+			ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("providerSpec.properties.storageProfile.osDisk.createOption")}))),
+		},
+		{
+			"should forbid zero osDisk size",
+			api.AzureOSDisk{Name: "osdisk-0", DiskSizeGB: 0, CreateOption: "Create"}, 1,
+			ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeInvalid), "Field": Equal("providerSpec.properties.storageProfile.osDisk.diskSizeGB")}))),
+		},
+		{
+			"should forbid negative osDisk size",
+			api.AzureOSDisk{Name: "osdisk-0", DiskSizeGB: -10, CreateOption: "Create"}, 1,
+			ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeInvalid), "Field": Equal("providerSpec.properties.storageProfile.osDisk.diskSizeGB")}))),
+		},
+	}
+
+	g := NewWithT(t)
+	for _, entry := range table {
+		t.Run(entry.description, func(t *testing.T) {
+			errList := validateOSDisk(entry.osDisk, fldPath)
+			g.Expect(len(errList)).To(Equal(entry.expectedErrors))
+			if entry.matcher != nil {
+				g.Expect(errList).To(entry.matcher)
+			}
+		})
+	}
+}
+
+func TestValidateOSProfile(t *testing.T) {
```

Comment:
we could either rename the function to mention which exact test case it is testing (as there is just being tested now) , or if we don't want that, then create a table inside?
```suggestion
func TestValidateNonEmptyAdminUsernameInOSProfile(t *testing.T) {
```

Reply:
The only mandatory field is AdminUserName and that is why adding table tests would really not add any value, do you think otherwise?